compile error for polodb-librocksdb-sys

Hi I am very interesting in poloDB and tried to write and then compile my first demo, but I found the compile error:

  cargo:warning=In file included from rocksdb/util/crc32c.cc:52:
  cargo:warning=rocksdb/util/crc32c.cc: In function 'uint64_t rocksdb::crc32c::CombineCRC(size_t, uint64_t, uint64_t, uint64_t, const uint64_t*)':
  cargo:warning=rocksdb/util/crc32c.cc:554:21: **error: '__builtin_ia32_pclmulqdq128' needs isa option -mpclmul -msse2**
  cargo:warning=  554 |   const auto res0 = _mm_clmulepi64_si128(crc0_xmm, multiplier, 0x00);
  cargo:warning=      |                     ^~~~~~~~~~~~~~~~~~~~
  cargo:warning=rocksdb/util/crc32c.cc:556:21: error: '__builtin_ia32_pclmulqdq128' needs isa option -mpclmul -msse2
  cargo:warning=  556 |   const auto res1 = _mm_clmulepi64_si128(crc1_xmm, multiplier, 0x10);
  cargo:warning=      |                     ^~~~~~~~~~~~~~~~~~~~


 --- stderr

  error occurred in cc-rs: command did not execute successfully (status code exit status: 1): LC_ALL="C" "c++" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-gdwarf-4" "-fno-omit-frame-pointer" "-m64" "-I" "rocksdb/include/" "-I" "rocksdb/" "-I" "rocksdb/third-party/gtest-1.8.1/fused-src/" "-I" "." "-Wall" "-Wextra" "-std=c++17" "-Wsign-compare" "-Wshadow" "-Wno-unused-parameter" "-Wno-unused-variable" "-Woverloaded-virtual" "-Wnon-virtual-dtor" "-Wno-missing-field-initializers" "-Wno-strict-aliasing" "-Wno-invalid-offsetof" "-msse2" "-std=c++17" "-DNDEBUG=1" "-DOS_LINUX" "-DROCKSDB_PLATFORM_POSIX" "-DROCKSDB_LIB_IO_POSIX" "-DROCKSDB_SUPPORT_THREAD_LOCAL" "-o" "/group/test_code/testpolodb/target/debug/build/polodb-librocksdb-sys-a75805157f38376d/out/3b2529d97c048fcf-crc32c.o" "-c" "rocksdb/util/crc32c.cc"


here is my gcc -v output:

[testpolodb]$ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/local/gcc/12.2.0_1/libexec/gcc/x86_64-redhat-linux/12.2.0/lto-wrapper
Target: x86_64-redhat-linux
Configured with: ../configure --prefix=/usr/local/gcc/12.2.0_1 --enable-bootstrap --enable-shared --enable-threads=posix --enable-tls --enable-checking=release --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --with-linker-hash-style=gnu --enable-languages=c,c++,fortran,lto --disable-libada --disable-libvtv --enable-plugin --enable-initfini-array --disable-libgcj --with-tune=corei7-avx --with-fpmath=avx --build=x86_64-redhat-linux --enable-targets=x86_64-redhat --target=x86_64-redhat-linux --host=x86_64-redhat-linux --disable-multiarch --disable-multilib --with-gmp=/home/xx/tickets/SCM-12326/build/ --with-mpfr=/home/xx/tickets/xx/build/ --with-mpc=/home/xx/tickets/xx/build/ --with-isl=/home/xx/tickets/xx/build/ CFLAGS='-O2 -gdwarf-4 -march=corei7-avx -mtune=corei7-avx -mavx -m64' CXXFLAGS='-O2 -gdwarf-4 -march=corei7-avx -mtune=corei7-avx -mavx -m64' FFLAGS='-O2 -gdwarf-4 -march=corei7-avx -mtune=corei7-avx -mavx -m64'
Thread model: posix
Supported LTO compression algorithms: zlib
gcc version 12.2.0 (GCC) 


can you help me with it, thank you very much

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -54,6 +54,28 @@ jobs:
       - name: Build Rust workspace
         run: cargo build --locked --release --workspace --exclude py-binding-polodb
 
+  rocksdb-x86-feature-compat:
+    name: RocksDB x86 Feature Compatibility
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v6
+        with:
+          submodules: recursive
+
+      - uses: actions/cache@v5
+        with:
+          path: |
+            ~/.cargo/bin/
+            ~/.cargo/registry/index/
+            ~/.cargo/registry/cache/
+            ~/.cargo/git/db/
+          key: ${{ runner.os }}-rocksdb-x86-feature-compat-${{ hashFiles('**/Cargo.lock') }}
+
+      - name: Build RocksDB with AVX but without PCLMUL
+        env:
+          CXXFLAGS: -mavx -mno-pclmul
+        run: cargo build --locked --release --verbose -p polodb-librocksdb-sys
+
   python-bindings:
     name: Python Bindings (${{ matrix.python-version }})
     runs-on: ubuntu-latest
@@ -142,6 +164,7 @@ jobs:
     if: ${{ always() }}
     needs:
       - rust-core
+      - rocksdb-x86-feature-compat
       - python-bindings
       - python-bindings-msrv
     runs-on: ubuntu-latest
@@ -149,11 +172,13 @@ jobs:
       - name: Verify required jobs succeeded
         env:
           RUST_CORE_RESULT: ${{ needs.rust-core.result }}
+          ROCKSDB_X86_FEATURE_COMPAT_RESULT: ${{ needs.rocksdb-x86-feature-compat.result }}
           PYTHON_BINDINGS_RESULT: ${{ needs.python-bindings.result }}
           PYTHON_BINDINGS_MSRV_RESULT: ${{ needs.python-bindings-msrv.result }}
         run: |
-          if [[ "$RUST_CORE_RESULT" != "success" || "$PYTHON_BINDINGS_RESULT" != "success" || "$PYTHON_BINDINGS_MSRV_RESULT" != "success" ]]; then
+          if [[ "$RUST_CORE_RESULT" != "success" || "$ROCKSDB_X86_FEATURE_COMPAT_RESULT" != "success" || "$PYTHON_BINDINGS_RESULT" != "success" || "$PYTHON_BINDINGS_MSRV_RESULT" != "success" ]]; then
             echo "Rust Core result: $RUST_CORE_RESULT"
+            echo "RocksDB x86 feature compatibility result: $ROCKSDB_X86_FEATURE_COMPAT_RESULT"
             echo "Python Bindings result: $PYTHON_BINDINGS_RESULT"
             echo "Python Bindings MSRV result: $PYTHON_BINDINGS_MSRV_RESULT"
             exit 1

--- a/src/librocksdb-sys/build.rs
+++ b/src/librocksdb-sys/build.rs
@@ -101,13 +101,11 @@ fn build_rocksdb() {
         .filter(|file| !matches!(*file, "util/build_version.cc"))
         .collect::<Vec<&'static str>>();
 
-    if let (true, Ok(target_feature_value)) = (
-        target.contains("x86_64"),
-        env::var("CARGO_CFG_TARGET_FEATURE"),
-    ) {
+    if target.contains("x86_64") {
         // This is needed to enable hardware CRC32C. Technically, SSE 4.2 is
         // only available since Intel Nehalem (about 2010) and AMD Bulldozer
         // (about 2011).
+        let target_feature_value = env::var("CARGO_CFG_TARGET_FEATURE").unwrap_or_default();
         let target_features: Vec<_> = target_feature_value.split(',').collect();
 
         if target_features.contains(&"sse2") {
@@ -132,6 +130,11 @@ fn build_rocksdb() {
         }
         if !target.contains("android") && target_features.contains(&"pclmulqdq") {
             config.flag_if_supported("-mpclmul");
+        } else {
+            // RocksDB 9 assumes AVX implies PCLMUL. That is not true for all
+            // x86-64 targets (or for custom CXXFLAGS), so explicitly disable
+            // its PCLMUL implementation unless Cargo enables the feature.
+            config.define("NO_PCLMUL", None);
         }
     }
 
@@ -265,6 +268,12 @@ fn build_rocksdb() {
     config.file("build_version.cc");
 
     config.cpp(true);
+    if !target.contains("windows") {
+        // Some RocksDB 9 headers use fixed-width integer types without
+        // including <cstdint> directly. Newer standard libraries no longer
+        // provide those declarations transitively in every translation unit.
+        config.flag("-include").flag("cstdint");
+    }
     config.flag_if_supported("-std=c++17");
     config.compile("librocksdb.a");
 }

